### PR TITLE
Create front end service, store and model for Notification Centre

### PR DESCRIPTION
**Pull request by rayking29**
# Implements [AB#14360](https://dev.azure.com/qslvic/304a1f8c-dace-4f85-adf3-bf563d5b3a39/_workitems/edit/14360)

## Description

Add Notification Centre to Health Gateway 
- create front end service, store and model

## Testing

- [ ] Unit Tests Updated
- [ ] Functional Tests Updated
- [x] Not Required

## Items to Review:

-   [General PR Guidelines](https://github.com/bcgov/healthgateway/wiki/PRguidance)

### CODE REVIEW

File: Apps/WebClient/src/ClientApp/src/services/restNotificationService.ts [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+                    return reject(
+                        ErrorTranslator.internalNetworkError(
+                            err,
+                            ServiceCode.HealthGatewayUser
+                        )
+                    );
+                });
+        });
+    }
+    deleteNotification(
```
Comment (BrianMaki) [inline Apps/WebClient/src/ClientApp/src/services/restNotificationService.ts:52]: Corresponding backend code has been named dismiss.  Was there a reason to not keep things consistent.

File: Apps/WebClient/src/ClientApp/src/services/restNotificationService.ts [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+                    return reject(
+                        ErrorTranslator.internalNetworkError(
+                            err,
+                            ServiceCode.HealthGatewayUser
+                        )
+                    );
+                })
+        );
+    }
+    deleteNotifications(hdid: string): Promise<void> {
```
Comment (BrianMaki) [inline Apps/WebClient/src/ClientApp/src/services/restNotificationService.ts:83]: Corresponding backend code has been named dismiss.  Was there a reason to not keep things consistent.

File: Apps/WebClient/src/ClientApp/src/store/modules/notification/actions.ts [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+                        context.dispatch("handleError", {
+                            error,
+                            errorType: ErrorType.Retrieve,
+                        });
+                        reject(error);
+                    });
+            }
+        });
+    },
+    deleteNotification(
```
Comment (BrianMaki) [inline Apps/WebClient/src/ClientApp/src/store/modules/notification/actions.ts:53]: Corresponding backend code has been named dismiss.  Was there a reason to not keep things consistent.

File: Apps/WebClient/src/ClientApp/src/store/modules/notification/actions.ts [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+                .catch((error: ResultError) => {
+                    context.dispatch("handleError", {
+                        error,
+                        errorType: ErrorType.Delete,
+                    });
+                    reject(error);
+                })
+        );
+    },
+    deleteNotifications(context, params: { hdid: string }): Promise<void> {
```
Comment (BrianMaki) [inline Apps/WebClient/src/ClientApp/src/store/modules/notification/actions.ts:77]: Corresponding backend code has been named dismiss.  Was there a reason to not keep things consistent.

File: Apps/WebClient/src/ClientApp/src/store/modules/notification/mutations.ts [hunk 5]
```diff
@@ -0,0 +1,24 @@
+import { ResultError } from "@/models/errors";
+import Notification from "@/models/notification";
+import { LoadStatus } from "@/models/storeOperations";
+
+import { NotificationMutations, NotificationState } from "./types";
+
+export const mutations: NotificationMutations = {
+    deleteNotification(state: NotificationState, notification: Notification) {
```
Comment (BrianMaki) [inline Apps/WebClient/src/ClientApp/src/store/modules/notification/mutations.ts:8]: Corresponding backend code has been named dismiss.  Was there a reason to not keep things consistent.

File: Apps/WebClient/src/ClientApp/src/store/modules/notification/mutations.ts [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+        const notificationIndex = state.notifications.findIndex(
+            (x) => x.id === notification.id
+        );
+        if (notificationIndex > -1) {
+            state.notifications.splice(notificationIndex, 1);
+        }
+    },
+    deleteNotifications(state: NotificationState) {
```
Comment (BrianMaki) [inline Apps/WebClient/src/ClientApp/src/store/modules/notification/mutations.ts:16]: Corresponding backend code has been named dismiss.  Was there a reason to not keep things consistent.

File: Apps/WebClient/src/ClientApp/src/store/modules/notification/types.ts [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+}
+
+type StoreContext = ActionContext<NotificationState, RootState>;
+export interface NotificationActions
+    extends ActionTree<NotificationState, RootState> {
+    retrieve(
+        context: StoreContext,
+        params: { hdid: string }
+    ): Promise<ApiResult<Notification[]>>;
+    deleteNotification(
```
Comment (BrianMaki) [inline Apps/WebClient/src/ClientApp/src/store/modules/notification/types.ts:36]: Corresponding backend code has been named dismiss.  Was there a reason to not keep things consistent.

File: Apps/WebClient/src/ClientApp/src/store/modules/notification/types.ts [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+        context: StoreContext,
+        params: { hdid: string; notification: Notification }
+    ): Promise<void>;
+    deleteNotifications(
```
Comment (BrianMaki) [inline Apps/WebClient/src/ClientApp/src/store/modules/notification/types.ts:40]: Corresponding backend code has been named dismiss.  Was there a reason to not keep things consistent.

File: Apps/WebClient/src/ClientApp/src/store/modules/notification/types.ts [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+        params: { error: ResultError; errorType: ErrorType }
+    ): void;
+}
+
+export interface NotificationMutations extends MutationTree<NotificationState> {
+    deleteNotification(
+        state: NotificationState,
+        notification: Notification
+    ): void;
+    deleteNotifications(state: NotificationState): void;
```
Comment (BrianMaki) [inline Apps/WebClient/src/ClientApp/src/store/modules/notification/types.ts:55]: Corresponding backend code has been named dismiss.  Was there a reason to not keep things consistent.

File: Apps/WebClient/src/ClientApp/src/services/restNotificationService.ts [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+    ): Promise<void> {
+        return new Promise((resolve, reject) =>
+            this.http
+                .delete<ApiResult<void>>(
+                    `${this.baseUri}${this.NOTIFICATION_BASE_URI}${hdid}/notificationId`,
+                    notification.id
+                )
+                .then((apiResult) => {
+                    this.logger.verbose(
+                        `deleteNotification result: ${JSON.stringify(
```
Comment (BrianMaki) [inline Apps/WebClient/src/ClientApp/src/services/restNotificationService.ts:64]: The controller for dismiss returns Task<IActionResult>. I don't think there is anything to stringify.  For example, in dependent controller delete, it returns ActionResult<RequestResult<DependentModel>> .  In that situation, DependentMddel would be something to stringify.

File: Apps/WebClient/src/ClientApp/src/services/restNotificationService.ts [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+    deleteNotifications(hdid: string): Promise<void> {
+        return new Promise((resolve, reject) =>
+            this.http
+                .delete<ApiResult<void>>(
+                    `${this.baseUri}${this.NOTIFICATION_BASE_URI}${hdid}`,
+                    hdid
+                )
+                .then((apiResult) => {
+                    this.logger.verbose(
+                        `deleteNotifications result: ${JSON.stringify(
```
Comment (BrianMaki) [inline Apps/WebClient/src/ClientApp/src/services/restNotificationService.ts:92]: The controller for dismiss returns Task<IActionResult>. I don't think there is anything to stringify.  For example, in dependent controller delete, it returns ActionResult<RequestResult<DependentModel>> .  In that situation, DependentMddel would be something to stringify.

**Review by sslaws — CHANGES_REQUESTED** [review]
Please review Brians feedback and update.

**Review by BrianMaki — APPROVED** [review]

File: Apps/WebClient/src/ClientApp/src/store/modules/notification/mutations.ts [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+import { NotificationMutations, NotificationState } from "./types";
+
+export const mutations: NotificationMutations = {
+    dismissNotification(state: NotificationState, notification: Notification) {
+        const notificationIndex = state.notifications.findIndex(
+            (x) => x.id === notification.id
+        );
+        if (notificationIndex > -1) {
+            state.notifications.splice(notificationIndex, 1);
+        }
```
Comment (MikeLyttle) [inline Apps/WebClient/src/ClientApp/src/store/modules/notification/mutations.ts:14]: Can be simplified:

```
state.notifications = state.notifications.filter(n => n.id != notification.id;
```

File: Apps/WebClient/src/ClientApp/src/store/modules/notification/mutations.ts [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+    },
+    dismissNotifications(state: NotificationState) {
+        state.notifications;
```
Comment (MikeLyttle) [inline Apps/WebClient/src/ClientApp/src/store/modules/notification/mutations.ts:17]: Should be:

```
 state.notifications = [];
```

File: Apps/WebClient/src/ClientApp/src/store/modules/notification/types.ts [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+import Notification from "@/models/notification";
+import { LoadStatus, Operation } from "@/models/storeOperations";
+import { RootState } from "@/store/types";
+
+export interface NotificationState {
+    notifications: Notification[];
+    statusMessage: string;
+    error?: ResultError;
+    status: LoadStatus;
+    lastOperation: Operation | null;
```
Comment (MikeLyttle) [inline Apps/WebClient/src/ClientApp/src/store/modules/notification/types.ts:21]: `lastOperation` is unused and should be removed.

File: Apps/WebClient/src/ClientApp/src/store/modules/notification/mutations.ts [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+    },
+    notificationError(state: NotificationState, error: ResultError) {
+        state.error = error;
+        state.statusMessage = error.resultMessage;
+        state.status = LoadStatus.ERROR;
+    },
+};
```
Comment (MikeLyttle) [inline Apps/WebClient/src/ClientApp/src/store/modules/notification/mutations.ts:24]: The mutations `setRequested` and `setNotifications` are missing.

**Review by MikeLyttle — CHANGES_REQUESTED** [review]

**Review by MikeLyttle — APPROVED** [review]

**Pull request merged**